### PR TITLE
libfs: add sync rwlock to leveldb storage

### PR DESCRIPTION
**Pull request by strib**
When leveldb syncs its storage, we need to prevent any other goroutines from making KBFS Write calls at the same time, because the sync fails if there's outstanding dirty data in the `folderBranchOps`.

This protects the sync by adding a read-write lock that's taken for reading whenever making modifications via the billy filesystem, and taken for writing whenever changes are being synced.

Issue: HOTPOT-1263

### CODE REVIEW

File: go/kbfs/libfs/leveldb_storage.go [hunk 1]
```diff
@@ -255,7 +267,19 @@ func (fs *levelDBStorage) sync() (err error) {
...
-func (fs *levelDBStorage) setMeta(fd storage.FileDesc) error {
+func (fs *levelDBStorage) sync() (err error) {
+	fs.syncLock.Lock()
+	defer fs.syncLock.Unlock()
+	return fs.syncLocked()
+}
+
+func (fs *levelDBStorage) syncRLocked() (err error) {
+	fs.syncLock.RUnlock()
+	defer fs.syncLock.RLock()
```
Comment (songgao) [inline go/kbfs/libfs/leveldb_storage.go]: The name of the method implies that when it's called it should already be locked while in fact it's locking inside. Maybe rename it?

Comment (songgao) [reply go/kbfs/libfs/leveldb_storage.go]: Same for `doLogRLocked` above. Perhaps `rlockAndSync` and `rlockAndDoLog`?

Comment (strib) [reply go/kbfs/libfs/leveldb_storage.go]: @songgao I think you have it backwards.  This code is actually _unlocking_  and _re-locking_ it, not locking it.  It assumes it's already read-locked before called, as the name implies.

Comment (strib) [reply go/kbfs/libfs/leveldb_storage.go]: I don't follow the point about `doLogRLocked()` -- that function doesn't do any locking and assumes the lock is already held.

Comment (songgao) [reply go/kbfs/libfs/leveldb_storage.go]: Ah you're right. I mis-read the `doLogRLocked` one because of the GH collapsing.

**Pull request merged**